File: dashboard/app/api/prometheus/schedule-cas-outcome/route.ts
/**
 * Copyright 2026 ScitiX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * GET /api/prometheus/schedule-cas-outcome
 *
 * Per-second rate of scheduler CAS (TriggerUpdateWithOptions) outcomes,
 * broken down by outcome label. Admin-only.
 *
 * Query params:
 *   cluster=<required>
 *   team=<optional>     user=<optional>     pool=<optional>
 *   preset=1h|6h|24h|7d  (default: 1h)  — OR —
 *   start=<unix>  end=<unix>
 *
 * Response: TimeSeriesData
 *   series[0] = "Success"
 *   series[1] = "Retriable"
 *   series[2] = "Hard"
 */

import { withPrometheusRoute, fetchPrometheusRange, rangeResultToSeries } from "../_shared"

export const GET = withPrometheusRoute(
  { auth: "admin", parseTime: "range" },
  async ({ config, sel, timeRange }) => {
    const { start, end, step, rateWindow } = timeRange
    const queries = [
      {
        name: "Success",
        query: `sum(rate(agentbox_schedule_cas_outcome_total{${sel},outcome="success"}[${rateWindow}]))`,
      },
      {
        name: "Retriable",
        query: `sum(rate(agentbox_schedule_cas_outcome_total{${sel},outcome="retriable"}[${rateWindow}]))`,
      },
      {
        name: "Hard",
        query: `sum(rate(agentbox_schedule_cas_outcome_total{${sel},outcome="hard"}[${rateWindow}]))`,
      },
    ]
    const results = await Promise.all(
      queries.map(({ query }) =>
        fetchPrometheusRange(config, query, start, end, step).catch(() => null),
      ),
    )
    const series = results.flatMap((result, i) =>
      result ? rangeResultToSeries(result, [queries[i].name]) : [],
    )
    return { series, _promql: queries.map((q) => q.query) }
  },
)

File: pkg/apiserver/service/envscheduler/route.go
// Copyright 2026 ScitiX
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package envscheduler

import (
	"context"
	"time"

	"k8s.io/apimachinery/pkg/types"

	agentsv1alpha1 "github.com/scitix/agent-sandbox/api/v1alpha1"
	"github.com/scitix/agent-sandbox/pkg/lifecycle/schedule"
)

// routeMulti handles the multi-member case. It builds one
// CandidateContext per local member, runs them through the scoring
// framework (filter → score → sort), then tries Enqueue on each pool
// in best-first order. The first scheduler whose Enqueue accepts wins.
// If every candidate refuses, the call returns RouteSaturated.
func (m *Manager) routeMulti(_ context.Context, envKey types.NamespacedName, entry *envEntry, req *schedule.ClaimRequest) RouteResult {
	cands := m.buildCandidates(envKey, entry)
	if len(cands) == 0 {
		return RouteResult{Kind: RouteNotFound}
	}
	ranked, _ := m.framework.Rank(cands)
	for _, c := range ranked {
		pool := m.pools.GetOrCreateScheduler(envKey.Namespace, c.Member.poolName, "", "")
		if pool == nil {
			continue
		}
		if pool.Enqueue(req) {
			return RouteResult{Kind: RouteLocal, Pool: pool}
		}
	}
	return RouteResult{Kind: RouteSaturated}
}

// buildCandidates materialises one CandidateContext per local member
// of the env. Reads PoolScheduler.Snapshot for live counters and
// SandboxEnv.Status.ObservedMember for DesiredReplicas + SaturatedUntil.
// Both are O(1) atomic / map lookups on the hot path.
func (m *Manager) buildCandidates(envKey types.NamespacedName, entry *envEntry) []CandidateContext {
	now := time.Now()
	// Pull observed-member projections + per-scaling-group sibling
	// totals from Env.Status (when available). The router has to know
	// DesiredReplicas to score Headroom, and Σ-siblings to apply the
	// group cap.
	desiredByMember := map[string]int32{}
	saturatedByMember := map[string]*time.Time{}
	if m.envGetter != nil {
		if env, ok := m.envGetter.GetEnv(envKey.Namespace, envKey.Name); ok {
			for _, c := range env.Status.Clusters {
				if c.ClusterID != m.local {
					continue
				}
				for i := range c.ObservedMembers {
					om := &c.ObservedMembers[i]
					desiredByMember[om.Name] = om.DesiredReplicas
					if om.SaturatedUntil != nil && om.SaturatedUntil.After(now) {
						t := om.SaturatedUntil.Time
						saturatedByMember[om.Name] = &t
					}
				}
			}
		}
	}

	// Σ DesiredReplicas per scalingGroup so each candidate's
	// "siblings desired" is just (groupSum - self).
	groupSum := map[string]int32{}
	for _, mr := range entry.members {
		if !mr.isLocal || mr.scalingGroup == "" {
			continue
		}
		groupSum[mr.scalingGroup] += desiredByMember[mr.poolName]
	}

	cands := make([]CandidateContext, 0, len(entry.members))
	for _, mr := range entry.members {
		if !mr.isLocal {
			continue // cross-cluster routing deferred
		}
		var snap schedule.Snapshot
		if sched := m.pools.GetScheduler(envKey.Namespace, mr.poolName); sched != nil {
			snap = sched.Snapshot()
		}
		desired := desiredByMember[mr.poolName]
		siblings := int32(0)
		if mr.scalingGroup != "" {
			siblings = groupSum[mr.scalingGroup] - desired
		}
		cands = append(cands, CandidateContext{
			Now:                now,
			Member:             mr,
			Snap:               snap,
			DesiredReplicas:    desired,
			SiblingsDesiredSum: siblings,
			SaturatedUntil:     saturatedByMember[mr.poolName],
		})
	}
	return cands
}

// unused but exported helper kept here for future cross-cluster work —
// gives callers a way to peek at the saturation map without exposing the
// internal Env.Status walk. Compiles in but is not called yet.
var _ = func() *agentsv1alpha1.SandboxEnv { return nil }
